File: python_system/python_shared/api/ad_status_stats.py
"""
Ad Status Statistics API
Returns counts of ads by status
Converted from PHP to Python
"""

from fastapi import APIRouter, Depends
from fastapi.responses import JSONResponse
from sqlalchemy.orm import Session
from sqlalchemy import func, case, desc
import time

import sys
from pathlib import Path
sys.path.insert(0, str(Path(__file__).parent.parent))

from models import Ad
from database import get_db

router = APIRouter()


@router.get("/ad_status_stats")
async def ad_status_stats(company: str = None, db: Session = Depends(get_db)):
    """Get ad status breakdown - matches PHP logic exactly"""
    try:
        # Build base query
        query = db.query(
            func.count(Ad.ad_id).label('total'),
            func.sum(case((Ad.status == 'active', 1), else_=0)).label('active'),
            func.sum(case((Ad.status == 'inactive', 1), else_=0)).label('inactive'),
            func.sum(case((Ad.status == 'scheduled', 1), else_=0)).label('scheduled'),
            func.sum(case((Ad.status == 'expired', 1), else_=0)).label('expired')
        )

        if company:
            query = query.filter(Ad.company_slug == company)

        stats = query.first()

        # Convert to integers (matching PHP behavior)
        result = {
            "total": int(stats.total or 0),
            "active": int(stats.active or 0),
            "inactive": int(stats.inactive or 0),
            "scheduled": int(stats.scheduled or 0),
            "expired": int(stats.expired or 0)
        }

        # Calculate percentages (matching PHP logic)
        total = result["total"] if result["total"] > 0 else 1
        result["percentages"] = {
            "active": round((result["active"] / total) * 100, 1),
            "inactive": round((result["inactive"] / total) * 100, 1),
            "scheduled": round((result["scheduled"] / total) * 100, 1),
            "expired": round((result["expired"] / total) * 100, 1)
        }

        # Get recent status changes (matching PHP query)
        recent_query = db.query(
            Ad.ad_id,
            Ad.status,
            Ad.updated_at
        )
        if company:
            recent_query = recent_query.filter(Ad.company_slug == company)

        recent_changes = recent_query.order_by(desc(Ad.updated_at)).limit(5).all()

        recent_changes_list = [
            {
                "ad_id": change.ad_id,
                "status": change.status,
                "updated_at": str(change.updated_at) if change.updated_at else None
            }
            for change in recent_changes
        ]

        return {
            "success": True,
            "stats": result,
            "recent_changes": recent_changes_list,
            "timestamp": int(time.time())
        }

    except Exception as e:
        return JSONResponse(
            status_code=500,
            content={
                "success": False,
                "error": "Failed to fetch ad status statistics",
                "message": str(e)
            }
        )

File: python_system/python_shared/api/ad_status_stats.php
<?php
/********************************************
 * Ad Status Statistics API
 * Returns counts of ads by status
 ********************************************/

header('Content-Type: application/json');
header('Access-Control-Allow-Origin: *');

require_once __DIR__ . '/../shared/database/Database.php';

$db = Database::getInstance();

try {
    // Get counts for each status
    $stats = $db->queryOne("
        SELECT
            COUNT(*) as total,
            SUM(CASE WHEN status = 'active' THEN 1 ELSE 0 END) as active,
            SUM(CASE WHEN status = 'inactive' THEN 1 ELSE 0 END) as inactive,
            SUM(CASE WHEN status = 'scheduled' THEN 1 ELSE 0 END) as scheduled,
            SUM(CASE WHEN status = 'expired' THEN 1 ELSE 0 END) as expired
        FROM ads
    ");

    // Convert to integers
    $result = [
        'total' => (int)($stats['total'] ?? 0),
        'active' => (int)($stats['active'] ?? 0),
        'inactive' => (int)($stats['inactive'] ?? 0),
        'scheduled' => (int)($stats['scheduled'] ?? 0),
        'expired' => (int)($stats['expired'] ?? 0)
    ];

    // Calculate percentages
    $total = $result['total'] ?: 1;
    $result['percentages'] = [
        'active' => round(($result['active'] / $total) * 100, 1),
        'inactive' => round(($result['inactive'] / $total) * 100, 1),
        'scheduled' => round(($result['scheduled'] / $total) * 100, 1),
        'expired' => round(($result['expired'] / $total) * 100, 1)
    ];

    // Get recent status changes
    $recentChanges = $db->query("
        SELECT
            ad_id,
            status,
            updated_at
        FROM ads
        ORDER BY updated_at DESC
        LIMIT 5
    ");

    echo json_encode([
        'success' => true,
        'stats' => $result,
        'recent_changes' => $recentChanges,
        'timestamp' => time()
    ], JSON_PRETTY_PRINT);

} catch (Exception $e) {
    http_response_code(500);
    echo json_encode([
        'success' => false,
        'error' => 'Failed to fetch ad status statistics',
        'message' => $e->getMessage()
    ], JSON_PRETTY_PRINT);
}

